File: src/torch_mpc/action_sampling/action_sampler.py
import torch

from torch_mpc.action_sampling.sampling_strategies.action_library import ActionLibrary
from torch_mpc.action_sampling.sampling_strategies.uniform_gaussian import UniformGaussian
from torch_mpc.action_sampling.sampling_strategies.gaussian_walk import GaussianWalk

class ActionSampler:
    """
    Main driver class for sampling actions
    """
    def __init__(self, sampling_strategies):
        self.sampling_strategies = sampling_strategies

        self.B = None
        self.K = 0
        self.H = None
        self.M = None
        self.device = None

        for k,strat in self.sampling_strategies.items():
            if self.B is None:
                self.B = strat.B
            else:
                assert self.B == strat.B, "Mismatch in B. Got {}, expected {}".format(self.B, strat.B)

            if self.H is None:
                self.H = strat.H
            else:
                assert self.H == strat.H, "Mismatch in H. Got {}, expected {}".format(self.H, strat.H)

            if self.M is None:
                self.M = strat.M
            else:
                assert self.M == strat.M, "Mismatch in M. Got {}, expected {}".format(self.M, strat.M)

            if self.device is None:
                self.device = strat.device
            else:
                assert self.device == strat.device, "Mismatch in device. Got {}, expected {}".format(self.device, strat.device)

            self.K += strat.K

    def sample_dict(self, u_nominal, u_lb, u_ub):
        samples = {k:v.sample(u_nominal, u_lb, u_ub) for k,v in self.sampling_strategies.items()}
        return samples

    def sample(self, u_nominal, u_lb, u_ub):
        samples = self.sample_dict(u_nominal, u_lb, u_ub)
        return torch.cat([v for v in samples.values()], dim=1)

    def to(self, device):
        self.device = device
        self.sampling_strategies = {k:v.to(device) for k,v in self.sampling_strategies.items()}
        return self

if __name__ == '__main__':
    import yaml
    import time
    import matplotlib.pyplot as plt

    config_fp = '/home/atv/physics_atv_ws/src/control/torch_mpc/configs/test_config.yaml'
    config = yaml.safe_load(open(config_fp, 'r'))
    device = config['common']['device']

    action_sampler = ActionSampler(config)

    u_nominal = torch.stack([
        torch.linspace(0., 1., 50),
        torch.linspace(-0.5, -0.2, 50)
    ], dim=-1).unsqueeze(0).to(device)

    u_lb = torch.tensor([0., -0.52], device=device).view(1, 2)
    u_ub = torch.tensor([1., 0.52], device=device).view(1, 2)

    t1 = time.time()
    samples = action_sampler.sample_dict(u_nominal, u_lb, u_ub)
    t2 = time.time()

    u_nominal = u_nominal.to('cpu')
    u_lb = u_lb.to('cpu')
    u_ub = u_ub.to('cpu')
    samples = {k:v.to('cpu') for k,v in samples.items()}

    print('took {:.4f}s to sample'.format(t2-t1))

    ## viz ##
    fig, axs = plt.subplots(2, 4, figsize=(24, 12))
    for i in range(axs.shape[0]):
        axs[i, 0].set_ylabel('act dim {}'.format(i))
        for j in range(axs.shape[1]):
            axs[i, j].set_ylim(u_lb[0, i]-0.1, u_ub[0, i]+0.1)
            axs[i, j].axhline(u_lb[0, i], color='k')
            axs[i, j].axhline(u_ub[0, i], color='k')

    colors = 'rgb'
    for i in range(axs.shape[0]):
        for j, (k, x) in enumerate(samples.items()):
            axs[i, j].plot(x[0, :, :, i].T, c=colors[j], alpha=0.1)
            axs[i, -1].plot(x[0, :, :, i].T, c=colors[j], alpha=0.1)
            axs[i, j].set_title(k)

    for i in range(axs.shape[0]):
        for j in range(axs.shape[1]):
            axs[i, j].plot(u_nominal[0, :, i], c='k', label='nominal')

    fig.suptitle('sampling (took {:.4f}s)'.format(t2-t1))
    plt.show() 

File: src/torch_mpc/action_sampling/action_sampler.cpp
#include "action_sampler.h"

int main()
{
    YAML::Node config = YAML::LoadFile("/home/anton/Desktop/SPRING24/AEC/torch_mpc/configs/costmap_speedmap.yaml");
    const std::string device_config = config["common"]["device"].as<std::string>();
    std::optional<torch::Device> device;

    if (device_config == "cuda")
    {
        device.emplace(torch::kCUDA, 0);
    }
    else if (device_config == "cpu")
    {
        device.emplace(torch::kCPU);
    }
    else
    {
        throw std::runtime_error("Unknown device " + device_config);
        return 1;
    }

    const int B = config["common"]["B"].as<int>();
    const int H = config["common"]["H"].as<int>();
    const int M = config["common"]["M"].as<int>();
    const int dt = config["common"]["dt"].as<int>();
    
    std::unordered_map<std::string, std::unique_ptr<SamplingStrategy>> sampling_strategies;

    for(auto iter = config["sampling_strategies"]["strategies"].begin(); iter != config["sampling_strategies"]["strategies"].end(); ++iter)
    {
        auto sv = iter->second;

        std::string type = sv["type"].as<std::string>();
        const std::vector<double> scale = sv["args"]["scale"].as<std::vector<double>>();
        if (type == "UniformGaussian")
        {
            const int K = sv["args"]["K"].as<int>();

            auto strategy = std::make_unique<UniformGaussian>(scale, B, K, H, M, *device);
            sampling_strategies.emplace(sv["label"].as<std::string>(), std::move(strategy));
        }
        else if (type == "ActionLibrary")
        {
            const int K = sv["args"]["K"].as<int>();
            std::string path = sv["args"]["path"].as<std::string>();

            auto strategy = std::make_unique<ActionLibrary>(path, B, K, H, M, *device);
            sampling_strategies.emplace(sv["label"].as<std::string>(), std::move(strategy));
        }
        else if (type == "GaussianWalk")
        {
            const int K = sv["args"]["K"].as<int>();

            std::unordered_map<std::string, std::variant<std::string, std::vector<double>>> initial_distribution;

            initial_distribution["type"] = sv["args"]["initial_distribution"]["type"].as<std::string>();
            initial_distribution["scale"] = sv["args"]["initial_distribution"]["scale"].as<std::vector<double>>();

            const std::vector<double> alpha = sv["args"]["alpha"].as<std::vector<double>>();

            auto strategy = std::make_unique<GaussianWalk>(initial_distribution, scale, alpha, B, K, H, M, *device);
            sampling_strategies.emplace(sv["label"].as<std::string>(), std::move(strategy));
        }
        else
        {
            throw std::runtime_error("Unknown sampling strategy " + type);
        }
    }

    ActionSampler action_sampler(sampling_strategies);
    
    const torch::Tensor u_nominal = torch::stack({
        torch::linspace(0.0, 1.0, 50),
        torch::linspace(-0.5, -0.2, 50)
    }, 1).unsqueeze(0);

    u_nominal.to(*device);

    const torch::Tensor u_lb = torch::tensor({0., -0.52}, *device).view({1, 2});
    const torch::Tensor u_ub = torch::tensor({1., 0.52}, *device).view({1, 2});

    auto t1 = std::chrono::high_resolution_clock::now();
    auto samples = action_sampler.sample_dict(u_nominal, u_lb, u_ub);
    auto t2 = std::chrono::high_resolution_clock::now();

    u_nominal.to(torch::kCPU);
    u_lb.to(torch::kCPU);
    u_ub.to(torch::kCPU);

    for(const auto& [k, v] : samples)
    {
        v.to(torch::kCPU);
    }

    std::cout << "took " << std::chrono::duration_cast<std::chrono::milliseconds>(t2 - t1).count()/1000 << "s to sample\n";

    return 0;
}